File: scripts/extract_motif_family.php
<?php
/**
 * extract_motif_family.php (PHP equivalent of the Python version).
 * Read FULL annotation file Plantpan3_TF_HQ_motif2species and emit
 * a SLIM mapping: matrix_id -> family + species list. Build-time
 * helper only; never run this inside the published image.
 *
 * Usage:
 *   php extract_motif_family.php <input_motif2species> <output_json>
 */

if ($argc < 3) {
    fwrite(STDERR, "Usage: php extract_motif_family.php <input> <output>\n");
    exit(1);
}
$in  = $argv[1];
$out = $argv[2];
if (!is_file($in)) { fwrite(STDERR, "ERROR: input not found: $in\n"); exit(1); }

$fc = []; $sc = [];
$fp = fopen($in, 'r');
while (($line = fgets($fp)) !== false) {
    $line = rtrim($line, "\r\n");
    if (strlen($line) < 2) continue;
    $cols = explode("\t", $line);
    if (count($cols) < 4) continue;
    $m  = trim($cols[0]);
    $sp = trim($cols[2] ?? '');
    $fa = trim($cols[3] ?? '');
    if ($m === '') continue;
    if (!isset($sc[$m])) $sc[$m] = [];
    if ($sp !== '') $sc[$m][$sp] = true;
    if ($fa !== '') {
        if (!isset($fc[$m])) $fc[$m] = [];
        $fc[$m][$fa] = ($fc[$m][$fa] ?? 0) + 1;
    }
}
fclose($fp);

$slim = [];
foreach (array_unique(array_merge(array_keys($fc), array_keys($sc))) as $m) {
    $fams = $fc[$m] ?? [];
    arsort($fams);
    $best = $fams ? array_key_first($fams) : '(uncharacterized)';
    $spc_list = array_keys($sc[$m] ?? []);
    sort($spc_list);
    $slim[$m] = [
        'family'        => $best,
        'species_count' => count($spc_list),
        'species'       => $spc_list,
    ];
}
ksort($slim);

$doc = [
    '_meta' => [
        'description' => 'PlantPAN5-offline slim motif metadata. matrix_id -> family + species list.',
        'count'       => count($slim),
    ],
    'motifs' => $slim,
];
file_put_contents($out, json_encode($doc, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES));
fprintf(STDERR, "Wrote %d motifs -> %s (%d bytes)\n", count($slim), $out, filesize($out));

File: scripts/extract_motif_family.py
#!/usr/bin/env python3
"""
Build-time helper: read the FULL annotation file
  Plantpan3_TF_HQ_motif2species
and emit a SLIM mapping of matrix_id -> family_name + species list.
Everything else (gene IDs, TF locus, ChIP-seq, PPI, homologs) is
intentionally dropped so the result is safe to ship inside a public
Docker image. The species list is bundled because it's already public
(JASPAR / CIS-BP / literature) and enables the species filter UI in
the offline result page.

Usage:
  python3 extract_motif_family.py <input_motif2species> <output_json>
"""
import json
import os
import sys

if len(sys.argv) < 3:
    sys.stderr.write(
        "Usage: extract_motif_family.py <input_motif2species> <output_json>\n")
    sys.exit(1)

src, out = sys.argv[1], sys.argv[2]
if not os.path.isfile(src):
    sys.stderr.write(f"ERROR: input not found: {src}\n"); sys.exit(1)

family_count = {}
species_count = {}

with open(src, errors="replace") as f:
    for line in f:
        line = line.rstrip("\r\n")
        if len(line) < 2: continue
        cols = line.split("\t")
        if len(cols) < 4: continue
        m  = cols[0].strip()
        sp = cols[2].strip() if len(cols) > 2 else ""
        fa = cols[3].strip() if len(cols) > 3 else ""
        if not m: continue
        species_count.setdefault(m, set())
        if sp: species_count[m].add(sp)
        if fa:
            family_count.setdefault(m, {})
            family_count[m][fa] = family_count[m].get(fa, 0) + 1

slim = {}
for m in sorted(set(list(family_count) + list(species_count))):
    fams = family_count.get(m, {})
    best = max(fams, key=fams.get) if fams else "(uncharacterized)"
    spc_list = sorted(species_count.get(m, set()))
    slim[m] = {
        "family": best,
        "species_count": len(spc_list),
        "species": spc_list,
    }

doc = {
    "_meta": {
        "description": "PlantPAN5-offline slim motif metadata. matrix_id -> family + species list. No gene IDs, no homologs, no ChIP-seq, no PPI.",
        "count": len(slim),
    },
    "motifs": slim,
}

os.makedirs(os.path.dirname(out) or ".", exist_ok=True)
with open(out, "w") as g:
    json.dump(doc, g, indent=2)
sys.stderr.write(f"Wrote {len(slim)} motifs -> {out} ({os.path.getsize(out)} bytes)\n")
